### ISSUE TITLE
JWT token and UUID parameters

### ISSUE THREAD

**Issue by mbiuki** [schemathesis/schemathesis#1125]
[Documentation](https://schemathesis.readthedocs.io/en/stable/) is not clear about the following matters:
### JWT token and UUID Headers
I know to some degree in previous issues this has been discussed but I am not very clear how I can pass in all the parameter UUIDs in my Python test function. Here is how my openapi spec is looking like:

```
  '/xyz/{xyzId}/something':
    post:
      summary: describe something
      description: Take a look at this snippet.
      x-ijk-docs-review: true
      operationId: takeSomeItem
      tags:
        - xyz
      parameters:
        - name: X-KLM-ID
          in: header
          description: KLM ID.
          required: true
          schema:
            type: string
            format: uuid
        - in: path
          name: xyzId
          description: xyz ID.
          x-soph-docs-review: true
          required: true
          schema:
            type: string
            format: uuid
```

- For header `X-KLM-ID` above as it is a UUID that must always accommodate with request headers, how can I construct a header as such with my tests, would the following be the optimized way:
```
import schemathesis
import pytest
from hypothesis import settings, Phase

URL="https://api-ca01.dev.local.pqr.com"

schema = schemathesis.from_path(
	"openapi.yaml",
	skip_deprecated_operations=True,
	validate_schema=True,
	base_url=URL
	)

@pytest.fixture
def token():
    return "<JWT token is generated every 720 seconds!>"

@schema.parametrize()
@settings(max_examples=25)
def test_api(case, token):
	case.headers = {"Authorization": f"Bearer {token}"}
	case.headers["X-KLM-ID"] = "d4dd94e9-dcb2-4e1c-927d-b257e5e73261"
	case.call_and_validate()
```
#### Questions to help quick start guide improvements
- Is the passage of headers as above correct? or there must be a different way to pass `X-KLM-ID` and `authorization token`?
- Where can I find all the options that can go with `schemathesis.from_path()`, for example, can I pass `--checks=all` the same way we do with CLI?
- I know there is no JWT token refreshable option. But, how can we auto-refresh the JWT token while the program is running? 
 Should we consider some auto-refresh timed window inside this function:
```
@pytest.fixture
def token():
    return "<JWT token is generated every 720 seconds!>"
```
Is there a way to have similar options to what RESTler carries? e.g., option parameter that can be described in user_settings.json, e.g.,
```
    "token_refresh_cmd": "~/refresh_token.py",
    "token_refresh_interval": 720,
```
### UUID path params
With RESTler, I can pass those fixed value uuid path parameters in the dictionary file. But, I wonder how can I do this in Schemathesis' python test file? I have about 10 UUID params. For example, in RESTler I can do this:
```
  "restler_custom_payload_header": {
    "X-KLM-ID": [
      "d4dd94e9-dcb2-4e1c-927d-b257e5e73261"
    ]
  },
  "restler_custom_payload": {
    "internalId": [
      "f4dd94e8-dcb2-4e8c-927d-b257e5e73232"
    ],
    "externalId": [
      "7dafa30d-25d9-4e1c-80b6-f842e4412f51"
    ],
    "homeId": [
      "316e513d-bcf1-474f-8273-8cfe96f9bd4f"
    ],
    "faceAgentId": [
      "316e513d-bcf1-474f-8273-8cfe96f8bd4f"
    ],
    "userId": [
      "308051ea-8724-41ab-b22b-46dda710e2a6"
    ],
    "groupId": [
      "8537e5dc-0359-4af9-b88c-dcaf7b26376e"
    ],
    "answerIds": [
      "b12b26e1-cb8c-4d72-8bdc-052aac79d981",
      "8dc3ade9-e1af-4c47-9309-350aab23517d"
    ]
  }
```
Is there any analogous way to make the above in my python tests for Schemethesis. i.e., is the following the right way:
```
@schema.parametrize()
@settings(max_examples=25)
def test_api(case, token):

	case.headers = {"Authorization": f"Bearer {token}"}
	case.headers["X-KLM-ID"] = "d4dd94e9-dcb2-4e1c-927d-b257e5e73261"

	case.path_parameters["internalId"]    = "f4dd94e8-dcb2-4e8c-927d-b257e5e73232"
	case.path_parameters["externalId"]  = ["...", "..."]
	case.path_parameters["..."] = "..."

	case.call_and_validate()

```

**Comment by mbiuki** [comment]
In line with the above note, I am trying to add the path parameters using `case.path_parameters` for {dataRegion}:
```
servers:
  - url: 'https://api-{dataRegion}.local.xyz.com'
    description: local service environment.
    x-ijk-docs-review: true
    variables:
      dataRegion:
        description: Data region where data stored.
        x-ijk-docs-review: true
        default: ca01
        enum:
          - ca01
          - ca02
          - ca01
          - ca02
          - ca03
```
But I keep getting the following error:
```
    case.path_parameters["dataRegion"]  = "ca01"
TypeError: 'NoneType' object does not support item assignment
```

I have a similar issue for the rest of the body path parameters as described above from OpenAPI for UUIDs.

This is how I have a python test:
```
@schema.parametrize()
@settings(max_examples=25)
def test_api(case, token):
	case.headers = {"Authorization": f"Bearer {token}"}
	case.headers["X-Tenant-ID"] = "d4dd94e9-dcb2-4e1c-927d-b257e5e73261"
        case.path_parameters["dataRegion"]  = "ca01"
	case.call_and_validate()
```

**Comment by Stranger6667** [comment]
Hi @mbiuki !

Thank you for asking :) I'll try to answer the questions below and will think about how they could be accommodated in the docs.

> For header X-KLM-ID above as it is a UUID that must always accommodate with request headers, how can I construct a header as such with my tests, would the following be the optimized way:

> Is the passage of headers as above correct? or there must be a different way to pass X-KLM-ID and authorization token?

I think that it depends on a particular test's goals and reflects how deep it reaches into the application code.

If the goal is to go beyond authorization logic and 404 responses on non-existing ID, then yes, for Python tests it is the way to go. Maybe a more generic version might look like this:

```python
@schema.parametrize()
def test_api(case, token):
    case.headers = case.headers or {}
    case.headers["Authorization"] = f"Bearer {token}"
    case.headers["X-KLM-ID"] = "d4dd94e9-dcb2-4e1c-927d-b257e5e73261"
```

In this case, the attribute itself is not overridden - if there are other headers, they won't be lost. But based on the schema you provided, your example should work :)

If the goal is to focus on the authorization code or code that handles this UUID value, then Schemathesis can generate them. In this particular case, `Authorization` header will be generated if the `securitySchemes` key defines it, and UUID requires [a custom string strategy](https://schemathesis.readthedocs.io/en/stable/extending.html#custom-string-strategies):

```python
from hypothesis import strategies as st
import schemathesis

schemathesis.register_string_format("uuid", st.uuids().map(str))
```

The UUID format is not included by default in Schemathesis (but probably it makes sense to have it this way, though) as it is not defined in the OAS spec.

> Where can I find all the options that can go with schemathesis.from_path(), for example, can I pass --checks=all the same way we do with CLI?

Unfortunately, they are quite underdocumented in [the public API reference](https://schemathesis.readthedocs.io/en/stable/api.html#loaders) and not much visible at the moment. It surely can be improved. 
The checks behavior is described in [this section](https://schemathesis.readthedocs.io/en/stable/python.html#how-are-responses-checked).

> I know there is no JWT token refreshable option. But, how can we auto-refresh the JWT token while the program is running?
Should we consider some auto-refresh timed window inside this function:
> Is there a way to have similar options to what RESTler carries?

There is [a relevant issue](https://github.com/schemathesis/schemathesis/issues/966 ) for that, but it is mostly related to CLI. For Python code some caching (e.g. from the [cachetools](https://github.com/tkem/cachetools/blob/master/cachetools/ttl.py#L55) library) might solve it:

```python
@pytest.fixture
def token_getter():

    @cached(cache=TTLCache(ttl=720))
    def retrieve_token():
        ...  # E.g. make a call to the API

    return retrieve_token

...
def test_api(case, token_getter):
    token = token_getter()
    case.headers = {"Authorization": f"Bearer {token}"}
    ...
```

As Hypothesis tests are executed many times within a single Pytest-level test, it is crucial to make a call within the test body. I will think about the problem more generally and maybe include some facilities to simplify auth handling.

> Is there any analogous way to make the above in my python tests for Schemethesis. i.e., is the following the right way:

It depends on the number of possible values for each parameter. If there is only one single value, then yes, you can set them directly. If you'd like to take a random value from the list of possible values, it should be accounted with the underlying Hypothesis machinery (otherwise, the test may be concluded as flaky). Here is an example:

```python
from hypothesis import strategies as st

@schema.parametrize()
@schema.given(data=st.data())
def test_password_reset(data, case):
    case.path_parameters["externalId"]  = data.draw(st.sampled_from(["a", "b", "c"]))
    ...
```

You may find more info in [this docs section](https://schemathesis.readthedocs.io/en/stable/python.html#using-additional-hypothesis-strategies) (Sphinx mangled all lists for some reason on readthedocs). Note that the `sampled_from` argument should be ordered.

> I have a similar issue for the rest of the body path parameters as described above from OpenAPI for UUIDs.

Interesting! Schemathesis does not consider variables in `servers` for data generation. It just takes the `default` which is required by the spec (so it always there) and interpolates the URL template. `path_parameters` are only for the [parameters](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#parameterObject)(`Parameter Object` in terms of Open API) that have `path` in their `in` field. But I never actually thought about deferring server URL template interpolation to later stages. I need to think about it more.

Similar behavior might happen if there no header parameters defined - Schemathesis will generate `None` for such cases. For `requestBody`, it might also generate a special sentinel value `NOT_SET` to distinguish cases when the body is not defined, and it has `None` as the value.

At the moment, I can suggest passing the `base_url` parameter to `call_and_validate` explicitly if you'd like to use `dataRegion` different from the default:

```
...
def test_api(case, token):
    ...
    case.call_and_validate(base_url="'https://api-ca02.local.xyz.com'")
```

There is no public API to get that URL template, but you can do it with something like `case.operation.schema.raw_schema.get("servers")[0]["url"]` depending on your actual `servers` value.

I hope that it helps :) Let me know what do you think! I'd be happy to elaborate and will plan to work on the documentation to reflect these points.

Cheers,
Dmitry

**Comment by mbiuki** [comment]
Hi @Stranger6667, 

Thanks for your response. I went halfway through your responses in the above, and first want to check a few things before moving on with the second half of your responses. Please see the enumerations as follows.

1. So, is this a correct Python test code:
```
from hypothesis import strategies as st
from hypothesis import settings, Phase
import schemathesis
import pytest

# Load the schema
pathToYAML="/path/openapi.yaml"
URL="https://xyz.com"
schema = schemathesis.from_path( pathToYAML,
	                            skip_deprecated_operations=True,
	                            validate_schema=True,
	                            code_sample_style="curl",
	                            base_url=URL)

@pytest.fixture
def token():
    return printToken() #<=== function call to extract the JWT token.

@schema.parametrize()
def test_api(case, token):
    case.headers = case.headers or {}
    case.headers["Authorization"] = f"Bearer {token}"
    case.headers["X-KLM-ID"] = xklmid
    schemathesis.register_string_format("uuid", st.uuids().map(str))
    case.call_and_validate()
```
I have a function for `token` that auto-generates JWT, and with the above, I would just run `pytest myfile.py`. Is this correct?

2. I see that [all checks are applied by default](https://schemathesis.readthedocs.io/en/stable/python.html#how-are-responses-checked). So we do not really have to do anything for activating checks in the above snippet.
3. By the way, it looks like the above test generates a new JWT token for every test, why is that? I can see from the logs that for every test that is reporting the failure, JWT token value is different!
4. How can I increase the response time as some of the failures are due to check if the response occurred in less than 500ms? I know two parameters from CLI for this, namely, `--max-response-time=5000` and `--hypothesis-deadline=5000`. How can I apply them in my Python test snippet above? I could not check any of these in the [API public reference guide](https://schemathesis.readthedocs.io/en/stable/api.html#public-api-reference).

**Comment by Stranger6667** [comment]
First of all, I'd like to thank you for asking all these questions, I really appreciate that :)

> So, is this a correct Python test code:

Generally yes, and here is a little thing I'd propose to change. 

`schemathesis.register_string_format("uuid", st.uuids().map(str))` mutates global state, and it is better to have it on the module level rather than in the test body (probably it worth mentioning in the docs as well).

> By the way, it looks like the above test generates a new JWT token for every test, why is that?

`schema.parametrize` behaves similarly to `pytest.mark.parametrize`, and if your API schema has, e.g., 5 operations, then the `schema.parametrize` decorator will create 5 different pytest-level tests. Each of them will have its own `token` fixture as its scope defaults to "function", i.e. `token` will be called 5 times, and each pytest-level test will have its own token. 

The next step is that Schemathesis converts each pytest-level test to Hypothesis test via `hypothesis.given` under the hood. Hypothesis executes this test many times and preserves pytest fixtures as is. Here is the workflow for some hypothetical API with multiple operations (e.g. `A`, `B`):

```
- test_api[A]. `token` is evaluated
  - test_api[A]-1. Hypothesis generates `case` for operation A
  - test_api[A]-2. Hypothesis generates `case` for operation A
  - ...
- test_api[B]. `token` is evaluated
  - test_api[B]-1. Hypothesis generates `case` for operation B
  - ...
```

- Pytest will report 2 tests - `test_api[A]` and `test_api[B]`
- `token` will be evaluated twice
- Hypothesis will generate `case` instances and pass it and `token` to the test body

The usual way to solve it in non-Hypothesis-based tests is to change the fixture scope to `module` or `session`:

```python
@pytest.fixture(scope="session")
def token():
    return printToken() #<=== function call to extract the JWT token.
```

The fixture above will be evaluated only once per test run, and all tests will use the same `token`.

However, the token may expire during the test run, and it will be nice to refresh it when it happens. As it can happen at an arbitrary moment that is not bound to any test phase, I think that the best way to solve it is to call a TTL-cached version of the `printToken` function inside `test_api`.

```python
TOKEN_LIFETIME = 720
THRESHOLD = 5

@pytest.fixture(scope="session")
def token_getter():

    @cached(cache=TTLCache(ttl=TOKEN_LIFETIME - THRESHOLD))
    def retrieve_token():
        return printToken()

    return retrieve_token

...
def test_api(case, token_getter):
    token = token_getter()
    case.headers = {"Authorization": f"Bearer {token}"}
    ...
```

This way, if the token is older than 715 seconds at the beginning of a test, then it will be refreshed. The `THRESHOLD` value is in place to cover a bit of delay between getting the token and the actual token validation on the server-side. Still, it actually depends on the token validation rules (`leeway` and such). 

> How can I increase the response time as some of the failures are due to check if the response occurred in less than 500ms?

You can use the `hypothesis.settings` decorator:

```python
from hypothesis import settings

...

@schema.parametrize()
@settings(deadline=None)
def test_api(case, token):
    ...
```

This will disable the deadline completely, but the test still may fail due to [Hypothesis health checks](https://hypothesis.readthedocs.io/en/latest/healthchecks.html) - if the schema is too complex, etc. You can disable them via the `suppress_health_check` argument to `settings`.

I want to note some semantic differences between `--max-response-time` and `--hypothesis-deadline`.

The former controls the API call time that is considered as a test failure. In a Python test, it is equivalent to:

```python
from datetime import timedelta

...
def test_api(case):
    response = case.call()
    assert response.elapsed < timedelta(milliseconds=5000), "Too slow!"
```

The latter controls the time for which the whole test should run to be considered as invalid and reported as an error. I.e., it is more like a health check to prevent too long test runs.

**Comment by mbiuki** [comment]
Thanks, @Stranger6667, for your reply. For the cache section in your comment [above](https://github.com/schemathesis/schemathesis/issues/1125#issuecomment-820261512), it is noteworthy to mention that the Python `cachetools` library should be installed to get it to work. And `maxsize` must also be passed as a parameter to `TTLCache`.

`pip3 install cachetools`

```
from cachetools import cached, LRUCache, TTLCache

TOKEN_LIFETIME = 720
THRESHOLD = 5 
...
@pytest.fixture(scope="session")
def token_getter():
    @cached( cache = TTLCache( maxsize = 1, ttl = TOKEN_LIFETIME - THRESHOLD))
    def retrieve_token():
        return printToken() #<=== function call to extract the JWT token.

    return retrieve_token
```

**Comment by mbiuki** [comment]
Hi @Stranger6667, here, I am reviewing the second half of your comments [above](https://github.com/schemathesis/schemathesis/issues/1125#issuecomment-818681612). 

> > I know there is no JWT token refreshable option. But, how can we auto-refresh the JWT token while the program is running?
> Should we consider some auto-refresh timed window inside this function:
> Is there a way to have similar options to what RESTler carries?
> 
> There is a relevant issue for that, but it is mostly related to CLI. For Python code some caching (e.g. from the cachetools library) might solve it:
> 
> @pytest.fixture
> def token_getter():
> 
>     @cached(cache=TTLCache(ttl=720))
>     def retrieve_token():
>         ...  # E.g. make a call to the API
> 
>     return retrieve_token
> 
> ...
> def test_api(case, token_getter):
>     token = token_getter()
>     case.headers = {"Authorization": f"Bearer {token}"}
>     ...
> As Hypothesis tests are executed many times within a single Pytest-level test, it is crucial to make a call within the test body. I will think about the problem more generally and maybe include some facilities to simplify auth handling.

For now, the above solution is working. Thanks for providing guidelines. But, yes, it would be great if, similar to [RESTler](https://github.com/microsoft/restler-fuzzer/blob/main/docs/user-guide/Fuzzing.md#how-to-invoke-restler-in-fuzz-lean-or-fuzz-mode), there could be a command option for a refreshable token in seconds to update the token bearer autonomously. 
For example,
```
 --token-refresh-cmd": "/path/token.py",
 --token-refresh-interval": 720,
```
`token.py` could generate something like:
```
{u'db74fa8e-911f-40d5-94aa-774bc22002fc': {}}
Authorization: Bearer <....>
```
Since I have some experience with RESTler, the second line is always selected (the first line is ignored and I do not know why it is even needed).

**Comment by mbiuki** [comment]
Hi @Stranger6667, regarding the rest of your responses [above](https://github.com/schemathesis/schemathesis/issues/1125#issuecomment-818681612).

> Interesting! Schemathesis does not consider variables in servers for data generation. It just takes the default which is required by the spec (so it always there) and interpolates the URL template. path_parameters are only for the parameters(Parameter Object in terms of Open API) that have path in their in field. But I never actually thought about deferring server URL template interpolation to later stages. I need to think about it more.

For path parameters in headers, could there be a similar design we have for custom headers, e.g., `Authorization`, `X-KLM-ID`? Could we have something implemented for:
```
servers:
  - url: 'https://test-{dataRegion}.tune.xyz.com/path/vx'
    description: Regional service in the test environment.
    x-ijk-docs-review: true
    variables:
      dataRegion:
        description: Data region where data played.
        x-ijk-docs-review: true
        default: ca01
        enum:
          - ca01
          - ca02
          - qa01
          - qa02
          - qa03
```
`dataRegion` in the above is a path parameter in the header that specifies the server location. It does not need to be fuzzed since the server location is one fixed assignment only. Could we have an analogous template to `path_parameters` but for header, e.g.:
`case.path_header_parameters["dataRegion"]  = "ca01"`

**Comment by mbiuki** [comment]
Hi @Stranger6667, 

In addition to the above discussions, I also have some questions for below:

1. For path parameters with fixed UUIDs, can we pass multiple of them as like below:
```python
from hypothesis import strategies as st
from hypothesis import settings, Phase
import schemathesis
import pytest
...
@settings(deadline=None) # disable the deadline completely.
@schema.parametrize()
@schema.given(data=st.data())
def test_api(data, case, token_getter):
    token = token_getter()
    case.headers = case.headers or {}
    case.headers["Authorization"] = f"Bearer {token}"
    case.headers["X-KLM-ID"] = xklmid
    case.path_parameters["someId"] = "7dafa30d-25d9-4e1c-80b6-f842e4412f51"
    case.path_parameters["nextId"] = "316e513d-bcf1-474f-8273-8cfe96f9bd4f"
    case.path_parameters["manageId"] = "316e513d-bcf1-474f-8273-8cfe96f9bd4f"
    case.path_parameters["torId"] = "308051ea-8724-41ab-b22b-46dda920e2a6"
    case.path_parameters["wayId"] = "8537e5dc-0359-4af9-b88c-dcaf7b46376e"
    case.path_parameters["fontId"] = "b12b26e1-cb8c-4d72-8bdc-052aac79d981"
    case.path_parameters["ids"]  = data.draw(st.sampled_from(["b12b26e1-cb8c-4d72-8bdc-052aac79d981", "8dc3ade9-e2af-4c47-9309-350aab23517d"]))
    case.path_parameters["groupId"] = "MSxFdmVudDo6RW5kcG9pbnQ6Ok5vdFByb3RlY3RlZCw1MTMs"
    case.path_parameters["lastId"] = "8537e5dc-0359-4af9-b88c-dcaf7b46376e"
    case.call_and_validate()
```
And all the `Id`s are path params in `openapi`, for example, `manageId`:
 ```
 '/management/{manageId}':
    get:
      tags:
        - Endpoints
      summary: Get endpoint details
      description: Get an endpoint based on manage ID.
      x-ijk-docs-review: true
      operationId: getEndpointById
      parameters:
        - name: X-KLM-ID
          in: header
          description: KLM ID.
          required: true
          schema:
            type: string
            format: uuid
        - name: manageId
          in: path
          description: Manage ID.
          x-ijk-docs-review: true
          required: true
          schema:
            type: string
            format: uuid
```

- Is the above correct or is there any better way of describing fixed UUIDs? 
- Would the above additions extend the fuzzer into finding more thorough bugs?

**Comment by Stranger6667** [comment]
Hi @mbiuki 

> dataRegion in the above is a path parameter in the header that specifies the server location. It does not need to be fuzzed since the server location is one fixed assignment only. Could we have an analogous template to path_parameters but for header, e.g.:
case.path_header_parameters["dataRegion"] = "ca01"

I think that it is a possible way to go; yep - I will experiment with the implementation, then I can provide more info on that matter :)

> Will the above not work?

Yes, it should work.

> Is the above correct or is there any better way of describing fixed UUIDs?

Yes, it is a valid approach. Depending on the test body size, you might want to move, split it, and make modifications via a hook:

```python
from hypothesis import strategies as st


def before_generate_path_parameters(context, strategy):

    @st.composite
    def fill_parameters(draw, path_parameters):
        path_parameters["ids"] = draw(st.sampled_from(["b12b26e1-cb8c-4d72-8bdc-052aac79d981", "8dc3ade9-e2af-4c47-9309-350aab23517d"]))
        ...  # more modifications
        return path_parameters

    return strategy.flatmap(fill_parameters)

@settings(deadline=None) # disable the deadline completely.
@schema.parametrize()
@schema.hooks.apply(before_generate_path_parameters)
def test_api(case, token_getter):
    ...
```

It makes those modifications a bit more reusable as you can attach hooks to different functions + do filtration of generated values.

> Would the above additions extend the fuzzer into finding more thorough bugs?

Usually, yes, as the input params will more likely pass validation. But it depends on the actual schema and app - passing explicit values is excluding some random values for inputs which may also have some positive effect in bug finding :) But generally, yes, such additions usually make the tests go deeper into the codebase.

> Are there any logs that we could check live while fuzzing is in action to check which requests are being sent out and what are their responses?

I eventually plan to implement #701, which will provide a more or less common output format, but meanwhile, you can add some hand-crafted logging to the test function:

```python
import structlog

log = structlog.get_logger()

def test_api(case, token_getter):
    ... # modifications go here
    response = case.call()
    log.info(path=case.path, method=case.method, ...)  # put all needed case/response parameters there
    case.validate_response(response)
```

The code above uses `structlog` installed via `pip install structlog colorama` (`colorama` provides a colored output style). By attaching various log processors ([docs](https://www.structlog.org/en/stable/processors.html)), you can either output these logs to stdout or some file or maybe even logstash or a similar tool, which seems quite flexible to me. 

Sorry for the delay; I hope that my comments above help :)

**Comment by Stranger6667** [comment]
Closing this as #966 is implemented. Once released, the docs will contain a separate section on auth. The parameters part will be a part of my future work on the docs.
